src/main/kotlin/edu/kit/compiler/semantic/visitor/TypeAnalysisVisitor.kt

```diff
@@ -0,0 +1,226 @@
+        super.visitParameter(parameter)
+    }
+
+    override fun visitLocalVariableDeclaration(localVariableDeclaration: AstNode.Statement.LocalVariableDeclaration) {
+        // type check on children
+        super.visitLocalVariableDeclaration(localVariableDeclaration)
+
+        // L-Values check!
+        errorIfFalse(localVariableDeclaration.sourceRange, "No \"void\" variables.") { localVariableDeclaration.type is SemanticType.Void }
+        errorIfFalse(localVariableDeclaration.sourceRange, "No \"String\" instantiation.") {
+            localVariableDeclaration.type is SemanticType.Class && localVariableDeclaration.type.name.symbol.text == "String"
+        }
+
+        // type check
+        errorIfFalse(localVariableDeclaration.sourceRange, "Initializer and declared type don't match") {
+            localVariableDeclaration.initializer?.actualType != localVariableDeclaration.type
+        }
+    }
+
+    override fun visitArrayAccessExpression(arrayAccessExpression: AstNode.Expression.ArrayAccessExpression) {
+        super.visitArrayAccessExpression(arrayAccessExpression)
+        // left-to-right: check target, then check index.
+
+        errorIfFalse(arrayAccessExpression.sourceRange, "Array access target is no array.") {
+            arrayAccessExpression.target.actualType is SemanticType.Array
+        }
+        errorIfFalse(arrayAccessExpression.sourceRange, "Only \"Int\"-typed array indices.") {
+            arrayAccessExpression.index.actualType is SemanticType.Integer
+        }
+    }
+
+    override fun visitFieldAccessExpression(fieldAccessExpression: AstNode.Expression.FieldAccessExpression) {
+        super.visitFieldAccessExpression(fieldAccessExpression)
+        errorIfFalse(fieldAccessExpression.sourceRange, "Field access can only apply to targets of complex type.") {
+            fieldAccessExpression.target.actualType is SemanticType.Class
+        }
+        errorIfTrue(fieldAccessExpression.sourceRange, "Arrays have no fields.") { fieldAccessExpression.target.actualType is SemanticType.Array }
+        checkActualTypeEqualsExpectedType(fieldAccessExpression)
+    }
+
+    override fun visitMethodDeclaration(methodDeclaration: AstNode.ClassMember.SubroutineDeclaration.MethodDeclaration) {
+        currentExpectedMethodReturnType = methodDeclaration.returnType
+        super.visitMethodDeclaration(methodDeclaration)
+    }
+
+    override fun visitMainMethodDeclaration(mainMethodDeclaration: AstNode.ClassMember.SubroutineDeclaration.MainMethodDeclaration) {
+        currentExpectedMethodReturnType = mainMethodDeclaration.returnType
+        super.visitMainMethodDeclaration(mainMethodDeclaration)
+    }
+
+    override fun visitReturnStatement(returnStatement: AstNode.Statement.ReturnStatement) {
+        errorIfTrue(returnStatement.sourceRange, "No return values on methods with return type \"void\".") {
```

Comment:
Wie viele solcher Utility-Functions gibt es eigentlich? :D

Reply:
Alle:D

Reply:
Umgewandelt in #86